gRPC e-book chapter 1 (#14468)
dotnet/docs: docs/architecture/grpc-for-wcf-developers/introduction.md

---
title: Introduction - gRPC for WCF Developers
description: Introduction
author: markrendle
ms.date: 09/02/2019
---

# Introduction

> Helping machines communicate with each other has been one of the primary preoccupations of the digital age. Determining the optimal remote communication mechanism that will suit the interoperability demands of the current infrastructure. As you can imagine, that mechanism changes as either the demands or the infrastructure evolves.

Reviewer: Yeah, I missed a push. Fixed.

> The release of .NET Core 3.0 marks a shift in the way that Microsoft delivers remote communication solutions to developers who want to deliver services across a range of platforms. .NET Core doesn't offer Windows Communication Foundation (WCF) out of the box but, with the release of version ASP.NET Core 3.0, it does provide built-in gRPC functionality.

Reviewer 2: I think either here, or somewhere else in this introduction it is worth mentioning that gRPC is a popular framework in the wider programming community.

Basically I want to get the message across that gRPC wasn't picked randomly to be WCF's successor but because gRPC has mindshare.

Reviewer: Good call, thanks.

During the 1990s, the emphasis had been mainly focused on improving internal networks using the same language and platforms. TCP/IP became the gold standard for this type of communication.

However, the focus soon shifted to how best to optimize communication across multiple platforms promoting a language agnostic approach. Service-oriented architecture (SOA) provided a structure for loosely coupling a broad collection of services that could be provided to an application.

The development of *Web Services* occurred once all major platforms could access the Internet, but they still couldn’t interact with each other. Web services have open standards and protocols including:

- XML to tag and code data;
- Simple Object Access Protocol (SOAP) to transfer data;
- Web Services Definition Language (WSDL) – describes and connects the web service to the client application; *and*
- Universal Description Discovery Integration (UDDI)- enables Web Services to be discoverable by other Services

> SOAP defines the rules by which distributed elements of an application can communicate with each other even if they are on different platforms. SOAP is based on XML, so it's human-readable. The sacrifice for making SOAP easily understood is size; SOAP messages are larger than messages in comparable protocols. SOAP was designed to break down monolithic applications into multi-component form without losing security or control. Therefore, WCF was designed to work with that kind of system unlike gRPC, which began as a distributed system.

Reviewer 3: ```suggestion
SOAP defines the rules by which distributed elements of an application can communicate with each other even if they are on different platforms. SOAP is based on XML, so it's human-readable. The sacrifice for making SOAP easily understood is size; SOAP messages are larger than messages in comparable protocols. SOAP was designed to break down monolithic applications into multi-component form without losing security or control. Therefore, WCF was designed to work with that kind of system unlike gRPC, which began as a distributed system. WCF addressed some of these limitations by developing and documenting proprietary extension protocols for the SOAP stack, but at the cost of lack of support from other platforms.
```

## About this guide

This guide covers the key features of gRPC.  The early chapters take a high level look at the main features of WCF and compares them to gRPC. It identifies where the are direct correlations between WCF and gRPC and also where gRPC offers an advantage. Where there is no correlation between WCF and gRPC or where gRPC is not able to offer an equivalent solution to WCF, this guide will suggest workarounds or places to go for further information.

Using a set of sample WCF applications, Chapter 5 is a deep dive look at a converting the main types of WCF service (simple request-reply, one-way, and streaming) to their equivalents in gRPC.

The final section of the book looks at how to get the best from gRPC in practice, including using additional tools like Docker containers or Kubernetes to leverage to the efficiency of gRPC, and a detailed look at the monitoring with logging, metrics and distributed tracing.

## Whom this guide is for

> This guide was written for developers working in .NET or .NET Core who have previously used WCF and who are seeking to migrate their applications to a modern RPC environment for .NET Core 3.0 and later versions. It is hoped that the guide will also be of use more generally for developers upgrading or considering upgrading to .NET Core 3.0 who want to use the built in gRPC tool.

Reviewer 4: .NET or .NET Core -> did you mean .NET Framework?

"It is hoped" sounds strange

---
title: Introduction - gRPC for WCF Developers
description: Introduction
author: markrendle
ms.date: 09/02/2019
---

# Introduction

Helping machines communicate with each other has been one of the primary preoccupations of the digital age. In particular, there is an ongoing effort to determine the optimal remote communication mechanism that will suit the interoperability demands of the current infrastructure. As you can imagine, that mechanism changes as either the demands or the infrastructure evolves.

The release of .NET Core 3.0 marks a shift in the way that Microsoft delivers remote communication solutions to developers who want to deliver services across a range of platforms. .NET Core doesn't offer Windows Communication Foundation (WCF) out of the box but, with the release of version ASP.NET Core 3.0, it does provide built-in gRPC functionality.

> gRPC is a hugely popular framework in the wider software community, used by developers across many programming languages for modern RPC scenarios. The community and the ecosystem are vibrant and active, with support for the gRPC protocol being added to infrastructure components like Kubernetes, service meshes, load balancers and more. These factors, as well as its performance, efficiency and cross-platform compatibility, make it the natural choice for a successor to WCF.

Reviewer 2: ```suggestion
gRPC is a popular framework in the wider software community, used by developers across many programming languages for modern RPC scenarios. The community and the ecosystem are vibrant and active, with support for the gRPC protocol being added to infrastructure components like Kubernetes, service meshes, load balancers and more. These factors, as well as its performance, efficiency and cross-platform compatibility, make gRPC a good choice for WCF apps moving to .NET Core.
```

I'm not sure if we should explicitly say gRPC is WCF's successor. That was more the *vibe* I wanted to get across 😄

> gRPC is a popular framework in the wider software community, used by developers across many programming languages for modern RPC scenarios. The community and the ecosystem are vibrant and active, with support for the gRPC protocol being added to infrastructure components like Kubernetes, service meshes, load balancers and more. These factors, as well as its performance, efficiency and cross-platform compatibility, make gRPC a natural choice for new apps and WCF apps moving to .NET Core.

Reviewer 2: ```suggestion
gRPC is a popular framework in the wider software community, used by developers across many programming languages for modern RPC scenarios. The community and the ecosystem are vibrant and active, with support for the gRPC protocol being added to infrastructure components like Kubernetes, service meshes, load balancers and more. These factors, as well as its performance, efficiency and cross-platform compatibility, makes gRPC a natural choice for new apps and WCF apps moving to .NET Core.
```

Reviewer: No, should be "make".
